Math/NumberTheory/Primes/Factorisation/QuadraticSieve.hs

```diff
@@ -5,182 +5,289 @@
-      pure matrix
+            -- This line has to be changed
+            nextDecompositionOfA = decompositionOfA
+
+        goSelfInitSieving previousSmoothNumbers ((b, c) : otherCoeffs) = runST $ do
+          let
+            f x = a * x * x + 2 * b * x + c
+            (sievingInterval, sievingLogInterval) = V.unzip $ generateLogInterval f m
+          sievingLogIntervalM <- V.thaw sievingLogInterval
+          smoothLogSieveM sievingLogIntervalM (zip factorBase squareRoots) a b c m
+          sievedLogInterval <- V.unsafeFreeze sievingLogIntervalM
+          let
+            newSmoothNumbers = V.toList $ findLogSmoothNumbers factorBase m h decompositionOfA b $ V.zip sievingInterval sievedLogInterval
+            smoothNumbers = SS.toList . SS.fromList $ previousSmoothNumbers ++ newSmoothNumbers
+            matrixSmoothNumbers
+              -- Minimise length of matrix
+              | numberOfConstraints < length mat = trace ("Matrix dimension: " ++ show (numberOfConstraints, length mat)) $ take (numberOfConstraints + 10) smoothNumbers
+              | otherwise                        = trace ("Matrix dimension: " ++ show (numberOfConstraints, length mat)) $ goSelfInitSieving smoothNumbers otherCoeffs
+              where
+                numberOfConstraints = S.size $ foldr (\col acc -> acc <> I.keysSet col) mempty mat
+                mat = trace ("Log filtering: " ++ show (V.length (V.filter (< h) sievedLogInterval), length newSmoothNumbers)) $ fmap snd smoothNumbers
+          pure matrixSmoothNumbers
 
-    indexedSmoothNumbers = goSieving [] startingPoint 1
+    sievingData = removeRows $ goSieving [] initialDecompositionOfA
+    matrix = trace ("Size of Matrix: " ++ show (length sievingData)) $ translate $ fmap (convertToSet . snd) sievingData
 
-    goSolving :: Int -> [(Integer, SignedPrimeIntSet)] -> [(Integer, Integer)]
-    goSolving seed sievingData = firstSquare `seq` secondSquare `seq` (firstSquare, secondSquare) : goSolving (seed + 1) sievingData
+    goSolving :: Int -> [(Integer, Integer)]
+    goSolving seed = firstSquare `seq` secondSquare `seq` (firstSquare, secondSquare) : goSolving (seed + 1)
       where
-        firstSquare = findFirstSquare n (V.fromList (fmap fst sievingData)) solution
-        secondSquare = findSecondSquare n (V.fromList (fmap (snd . second primeSet) sievingData)) solution
-        solution = convertToList $ linearSolve' seed $ translate $ fmap snd sievingData
+        firstSquare = findFirstSquare n (fmap fst squaresData)
+        secondSquare = findSecondSquare n (fmap snd squaresData)
+        -- Add factorisation of a
+        squaresData = map (sievingData !!) solution
+        solution = convertToList $ linearSolve' seed matrix
 
-  pure $ goSolving (integerToInt n) indexedSmoothNumbers
+  pure $ goSolving (integerToInt n)
+
+generatePrimes :: Integer -> Double -> Int -> [Prime Int]
+generatePrimes n midPoint len = lowerPrimes ++ higherPrimes
```

Comment:
There is an simpler way to do it, if we can sacrifice some precision. Use [`approxPrimeCount`](http://hackage.haskell.org/package/arithmoi-0.11.0.0/docs/Math-NumberTheory-Primes-Counting.html#v:approxPrimeCount) and [`nthPrimeApprox`](http://hackage.haskell.org/package/arithmoi-0.11.0.0/docs/Math-NumberTheory-Primes-Counting.html#v:nthPrimeApprox) to estimate the lower bound (assuming that half of primes are positive residues). Then generate `len` primes starting from this lower bound.